pkg/util/validation/stack.go

```diff
@@ -28,7 +28,17 @@ func validateStackName(s string) error {
 	if stackNameRE.MatchString(s) {
 		return nil
 	}
-	return errors.New("a stack name may only contain alphanumeric, hyphens, underscores, or periods")
+	return errors.New("stack names may only contain alphanumeric, hyphens, underscores, or periods")
+}
+
+// validateStackTagName checks if s is a valid stack tag name, otherwise returns a descriptive error.
+// This should match the stack naming rules enforced by the Pulumi Service.
+func validateStackTagName(s string) error {
+	tagNameRE := regexp.MustCompile("^[a-zA-Z0-9-_.:]{1,40}$")
```

Comment:
We can move this out of the function -- I believe this makes sure the regex compiles the first time the package is loaded & then only has to compile the regex once.

Reply:
I thought about mentioning the same thing, but erred on the side of not mentioning it.

You are correct that `MustCompile` does the hard work of compiling the regex so it is faster to execute multiple times, and so ideally we'd only want to do that once. But the tradeoff is that calling `MustCompile` in, say, the package's `init()` function means that we _always_ do that work on program startup even if the `validateStackTagName` function is never called.

So I think the tradeoff here is "make `validateStackTagName` ever so much slower every time, vs. make every `pulumi` invocation ever so much slower for all commands.

The time we are talking about is definite going to be imperceptible to humans, so it isn't worth worrying about either way.